### PR TITLE
ON HOLD - Govee H5055 temp/battery only

### PR DESCRIPTION
**Pull request by DigiH**
As discussed in

https://github.com/theengs/decoder/issues/92

## Checklist:
  - [x] The pull request is done against the latest development branch
  - [x] Only one feature/fix was added per PR and the code change compiles without warnings
  - [x] I accept the [DCO](https://github.com/theengs/decoder/blob/development/docs/participate/development.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by h2zero** [comment]
I tested this with branch https://github.com/theengs/decoder/tree/check-prop-cond-func and it looks like it's working but the test fails due to incomplete expected data:
```
[ctest] Expected: {"brand":"Govee","model":"BluetoothBBQThermometer","model_id":"H5055","tempc1":23,"tempf1":73.4,"tempc2":115,"tempf2":239,"batt":70}
[ctest] Got JSON: {"brand":"Govee","model":"BluetoothBBQThermometer","model_id":"H5055","tempc1":23,"tempf1":73.4,"tempc2":115,"tempf2":239,"tempc3":23,"tempf3":73.4,"tempc4":115,"tempf4":239,"tempc5":23,"tempf5":73.4,"tempc6":115,"tempf6":239,"batt":70}
```

I'm unsure if the data retrieved is correct and the expected data should be changed? Looks reasonable to me though.

**Comment by DigiH** [comment]
There should only ever be 2 temps published per message, with three different subsequent messages having 2 different probe temps accordingly. I will look into this when finished my other tests with your new branch, @h2zero. Thanks.

**Comment by DigiH** [comment]
@h2zero Sill failing, but the included test cases all expect the correct results, for each recognition/publishing ONLY TWO temperatures reported. The problem here seems to be that the NOT function is not working if there are multiple OR conditions before. Not sure if the 'nested' condition checking was ever properly implemented into the property conditions as well, or only for the model condition - also with the new decoder functionality restructuring, as you tested before and me as well with the just now failing results, too.

We don't seem to have other such cases (yet), and this H5055 decoder also still has the initial model coindition serviceuuid issue for correct recognition.

**Comment by h2zero** [comment]
@DigiH I have updated the branch to support nesting arrays in the conditions.
It is working now with this PR using the following decoder:
```
R""""(
{
   "brand":"Govee",
   "model":"Bluetooth BBQ Thermometer",
   "model_id":"H5055",
   "condition":["manufacturerdata", ">=", 41, "index", 12, "06", "&", "manufacturerdata", "index", 26, "06", "&", "manufacturerdata", "index", 40, "0"],
   "properties":{
      "tempc1":{
        "condition":[["manufacturerdata", 10, "0", "|", "manufacturerdata", 10, "1", "|", "manufacturerdata", 10, "2", "|", "manufacturerdata", 10, "3"], "&", "manufacturerdata", 14, "!", "ffff"],
        "decoder":["value_from_hex_data", "manufacturerdata", 14, 4, true, false]
      },
      "tempc2":{
        "condition":[["manufacturerdata", 10, "0", "|", "manufacturerdata", 10, "1", "|", "manufacturerdata", 10, "2", "|", "manufacturerdata", 10, "3"], "&", "manufacturerdata", 28, "!", "ffff"],
        "decoder":["value_from_hex_data", "manufacturerdata", 28, 4, true, false]
      },
      "tempc3":{
        "condition":[["manufacturerdata", 10, "4", "|", "manufacturerdata", 10, "5", "|", "manufacturerdata", 10, "6", "|", "manufacturerdata", 10, "7"], "&", "manufacturerdata", 14, "!", "ffff"],
        "decoder":["value_from_hex_data", "manufacturerdata", 14, 4, true, false]
      },
      "tempc4":{
        "condition":[["manufacturerdata", 10, "4", "|", "manufacturerdata", 10, "5", "|", "manufacturerdata", 10, "6", "|", "manufacturerdata", 10, "7"], "&", "manufacturerdata", 28, "!", "ffff"],
        "decoder":["value_from_hex_data", "manufacturerdata", 28, 4, true, false]
      },
      "tempc5":{
        "condition":[["manufacturerdata", 10, "8", "|", "manufacturerdata", 10, "9", "|", "manufacturerdata", 10, "a", "|", "manufacturerdata", 10, "b"], "&", "manufacturerdata", 14, "!", "ffff"],
        "decoder":["value_from_hex_data", "manufacturerdata", 14, 4, true, false]
      },
      "tempc6":{
        "condition":[["manufacturerdata", 10, "8", "|", "manufacturerdata", 10, "9", "|", "manufacturerdata", 10, "a", "|", "manufacturerdata", 10, "b"], "&", "manufacturerdata", 28, "!", "ffff"],
        "decoder":["value_from_hex_data", "manufacturerdata", 28, 4, true, false]
      },
      "batt":{
        "decoder":["value_from_hex_data", "manufacturerdata", 8, 2, false]
      }
   }
})"""";
``` 

I had to adjust the expected result for one test as the expected temp values were incorrect:
`"{\"brand\":\"Govee\",\"model\":\"Bluetooth BBQ Thermometer\",\"model_id\":\"H5055\",\"tempc6\":85,\"tempf6\":185,\"batt\":70}",`

Changed to:
`"{\"brand\":\"Govee\",\"model\":\"Bluetooth BBQ Thermometer\",\"model_id\":\"H5055\",\"tempc6\":84,\"tempf6\":183.2,\"batt\":70}",`

**Comment by DigiH** [comment]
@h2zero still the same issue with the last & NOT not being properly recognised, therefor evaluating all 6 temperatures

```
1: Key:value count mismatch, result 16, expected 8
1: Expected: {"brand":"Govee","model":"Bluetooth BBQ Thermometer","model_id":"H5055","tempc1":23,"tempf1":73.4,"tempc2":115,"tempf2":239,"batt":70}
1: Got JSON: {"brand":"Govee","model":"BluetoothBBQThermometer","model_id":"H5055","tempc1":23,"tempf1":73.4,"tempc2":115,"tempf2":239,"tempc3":23,"tempf3":73.4,"tempc4":115,"tempf4":239,"tempc5":23,"tempf5":73.4,"tempc6":115,"tempf6":239,"batt":70}
```

**Comment by h2zero** [comment]
I think you missed the decoder json changes.

**Comment by DigiH** [comment]
Apologies, that went again when I corrected the one temperature you pointed out above. Looks all updated to me now but still failing. Or is still anything missing?

File: src/devices/H5055_json.h [hunk 1]
```diff
@@ -0,0 +1,72 @@
...nufacturerdata\",10,\"4\",\"|\",\"manufacturerdata\",10,\"5\",\"|\",\"manufacturerdata\",10,\"6\",\"|\",\"manufacturerdata\",10,\"7\",\"&\",\"manufacturerdata\",14,\"!\",\"ffff\"],\"decoder\":[\"value_from_hex_data\",\"manufacturerdata\",14,4,true,false]},\"tempc4\":{\"condition\":[\"manufacturerdata\",10,\"4\",\"|\",\"manufacturerdata\",10,\"5\",\"|\",\"manufacturerdata\",10,\"6\",\"|\",\"manufacturerdata\",10,\"7\",\"&\",\"manufacturerdata\",28,\"!\",\"ffff\"],\"decoder\":[\"value_from_hex_data\",\"manufacturerdata\",28,4,true,false]},\"tempc5\":{\"condition\":[\"manufacturerdata\",10,\"8\",\"|\",\"manufacturerdata\",10,\"9\",\"|\",\"manufacturerdata\",10,\"a\",\"|\",\"manufacturerdata\",10,\"b\",\"&\",\"manufacturerdata\",14,\"!\",\"ffff\"],\"decoder\":[\"value_from_hex_data\",\"manufacturerdata\",14,4,true,false]},\"tempc6\":{\"condition\":[\"manufacturerdata\",10,\"8\",\"|\",\"manufacturerdata\",10,\"9\",\"|\",\"manufacturerdata\",10,\"a\",\"|\",\"manufacturerdata\",10,\"b\",\"&\",\"manufacturerdata\",28,\"!\",\"ffff\"],\"decoder\":[\"value_from_hex_data\",\"manufacturerdata\",28,4,true,false]},\"batt\":{\"decoder\":[\"value_from_hex_data\",\"manufacturerdata\",8,2,false]}}}";
```
Comment (h2zero) [inline src/devices/H5055_json.h]: Still missing the changes :) `"condition":[["manufacturerdata"  .....  ], "&", "manufacturerdata"`

Comment (DigiH) [reply src/devices/H5055_json.h]: DUH!! :( I really need to clean up this branch, but something is still amiss.

**Comment by h2zero** [comment]
I meant that you were missing the extra braces for the nested conditions, sorry for not being clear, hard to be verbose while at work.

Try this:
[H5055_json.h.txt](https://github.com/theengs/decoder/files/8325557/H5055_json.h.txt)

**Comment by DigiH** [comment]
Great; Thanks !- nested conditions with brackets, makes the nesting a lot clearer - for the property conditions only, or also now applicable to the previously non-bracketed nested model condition?

**Comment by h2zero** [comment]
>or also now applicable to the previously non-bracketed nested model condition?

That's still a work in progress, this was the proof of concept.

**Comment by DigiH** [comment]
> That's still a work in progress, this was the proof of concept.

Thanks a lot! With it working here now in the property conditions, and the model condition working correctly as is already, and your other enhancements in the **check-prop-cond-func** branch, it should probably be best merged, until further refinement on this is being applied. Or what is your thought on this?

**Comment by h2zero** [comment]
It was a quick and dirty bit of code, needs some TLC, then I'll PR when ready.

**Comment by h2zero** [comment]
@DigiH I've cleaned up the code and made allowance for having conditions with NOT parameters to be placed in more efficient locations.

For instance this now works (no nested array):
```
 "tempc6":{
        "condition":["manufacturerdata", 28, "!", "ffff", "&", "manufacturerdata", 10, "8", "|", "manufacturerdata", 10, "9", "|", "manufacturerdata", 10, "a", "|", "manufacturerdata", 10, "b"],
        "decoder":["value_from_hex_data", "manufacturerdata", 28, 4, true, false]
      }
```

This will now skip recursive calling of the nested array if the NOT condition fails, saves a few cycles and works without nesting.

**Comment by h2zero** [comment]
@DigiH do you use vscode?

If so you could just install the cmake plugin and run the tests locally. Just a thought as it might save you some time 😀

**Comment by DigiH** [comment]
Ok, tempc5 is failing the NOT confition

```
1: Key:value count mismatch, result 8, expected 6
1: Expected: {"brand":"Govee","model":"Bluetooth BBQ Thermometer","model_id":"H5055","tempc6":84,"tempf6":183.2,"batt":70}
1: Got JSON: {"brand":"Govee","model":"Bluetooth BBQ Thermometer","model_id":"H5055","tempc5":65535,"tempf5":117995,"tempc6":84,"tempf6":183.2,"batt":70}
```

**Comment by DigiH** [comment]
> If so you could just install the cmake plugin and run the tests locally. Just a thought as it might save you some time 😀

I do and I will, thanks for the tip :) - just had a quick look - any particular CMake extension? 

Also double checking the decoder changes and cannot see any issues with

```
"tempc5":{
        "condition":["manufacturerdata", 14, "!", "ffff", "&", "manufacturerdata", 10, "8", "|", "manufacturerdata", 10, "9", "|", "manufacturerdata", 10, "a", "|", "manufacturerdata", 10, "b"],
        "decoder":["value_from_hex_data", "manufacturerdata", 14, 4, true, false]
      }, 
```

**Comment by h2zero** [comment]
Lol I had a feeling something like this would come up if I didn't test all of them 😄. Easy fix, updating now.

**Comment by DigiH** [comment]
:) all clear now, and definitely worth it with ignoring extra nested OR cycles, thanks!

**Comment by h2zero** [comment]
>just had a quick look - any particular CMake extension?

Cmake from twxs, Microsoft Cmake tools and C/C++ extensions should be enough.

**Comment by DigiH** [comment]
Thanks, got all those now, still a bit stuck with the tests not running, think might have to fiddle with the cmake settings on macOS a bit. Will try more later today.

**Comment by DigiH** [comment]
Hi @1technophile @h2zero I'm wondering if we should include this Govee H5055 BBQ Thermometer as is? Even with it not being sold any longer it seems to have been a popular BBQ item with good reviews and still traded on ebay. While the problem with the ServiceUUID array might still be an issue for other devices, I think the current restrictive multi-AND model condition does a good job of keeping the decoder unique.

I had to clean up all the files and merges, so created a new cleaner branch

https://github.com/DigiH/decoder/commits/H5055-new

@h2zero I kept the test log changes as a separate commit, as I wasn't sure if you just wanted that for the H5055 testing or if it should be included generally.

Thanks

**Comment by 1technophile** [comment]
@DigiH agreed. We can integrate it as is in my point of view.

**Comment by DigiH** [comment]
Closing this PR, and waiting for a decision about the test case logging changes before making a PR for the new branch.

Thanks

**Comment by h2zero** [comment]
I like the logging changes, I'd say leave them in.

**Pull request closed without merging**